### DISCUSSION TITLE
erro distribuição

### DISCUSSION THREAD

**Discussion by diogoheck6** [TadaSoftware/PyNFe#204] [General]
ao tentar acessar a opção 1 do dist NSU, me aparece o erro abaixo:

C:\Users\diogo.rodrigues\AppData\Local\Programs\Python\Python310\lib\site-packages\urllib3\connectionpool.py:1043: InsecureRequestWarning: Unverified HTTPS request is being made to host 'hom.nfe.fazenda.gov.br'. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/1.26.x/advanced-usage.html#ssl-warnings
  warnings.warn(

**Comment by diogoheck6** [comment]
tentei com outro codigo que peguei aqui para consulta de distribuição mas da o erro :
![image](https://user-images.githubusercontent.com/62776744/182467917-6ea5ec8e-9525-4a09-85a7-779b3d47ff45.png)

**Reply by leogregianin** [reply]
> tentei com outro codigo que peguei aqui para consulta de distribuição mas da o erro : ![image](https://user-images.githubusercontent.com/62776744/182467917-6ea5ec8e-9525-4a09-85a7-779b3d47ff45.png)

Algo estranho nesse seu código, está mostrando uma exceção na linha 28 porém, no seu código, a linha 28 está em branco.

**Comment by leogregianin** [comment]
> ao tentar acessar a opção 1 do dist NSU, me aparece o erro abaixo:
> 
> C:\Users\diogo.rodrigues\AppData\Local\Programs\Python\Python310\lib\site-packages\urllib3\connectionpool.py:1043: InsecureRequestWarning: Unverified HTTPS request is being made to host 'hom.nfe.fazenda.gov.br'. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/1.26.x/advanced-usage.html#ssl-warnings warnings.warn(

@diogoheck6 essa mensagem é um `warning` da lib urllib3 e não um erro, ou seja, mesmo com o warning a rotina funciona da mesma forma.

**Comment by diogoheck6** [comment]
Pois é Leo..eu copiei este codigo do git de vocês para tentar conseguir baixar os XML de entrada e saida de NFE e NFCE, pena não ter uma rotina para os de CTE. Pois deixa eu te explicar...hoje eu pego e baixo os extratos do site manualmente e largo numa pasta onde ai sim aplico uma rotina que faz com que se cruze as informações dos extratos com o SPED, gerando um excel com varias informações de quais notas estão no sped e nao estao nos extratos, etc...e Seria um grande saldo se eu conseguisse pegar essas informaçoes via webservice...porém nao entendo quase nada de webservice...até fui olhar as notas técnicas do sefaz para saber por onde começar mas não estou conseguindo..eu programo bastante em python...vou mostrar esse codigo inteiro que te mandei só um pedaço ..se tu puder tentar identificar esse erro..também achei estranho que ele aponta a linha 28 .. O certificado vou colocar como xxxxx por privacidade do cliente..LGPD..se tu puder me dar uma ajuda..ficaria muito feliz mesmo...


from pynfe.processamento.comunicacao import ComunicacaoSefaz
from pynfe.utils.descompactar import DescompactaGzip
from pynfe.utils.flags import NAMESPACE_NFE
from lxml import etree


certificado = 'xxxxxxxxxxxxx.pfx'
senha = '146531'
uf = 'rs'
homologacao = True

CNPJ = ''
CHAVE = ''  # deixar a chave de acesso vazia
NSU = 0

con = ComunicacaoSefaz(uf, certificado, senha, homologacao)
ultNSU = 0
maxNSU = 0
cStat = 0

while True:
    xml = con.consulta_distribuicao(cnpj=CNPJ, chave=CHAVE, nsu=NSU)
    NSU = str(NSU).zfill(15)
    print(f'Nova consulta a partir do NSU: {NSU}')

    with open(f'./xml/consulta_distrib_gzip-{NSU}.xml', 'w+') as f:
        f.write(xml.text)

    # resposta = etree.fromstring(xml.content)
    # print(resposta)
    resposta = etree.parse(f'./xml/consulta_distrib_gzip-{NSU}.xml')
    ns = {'ns': NAMESPACE_NFE}

    contador_resposta = len(resposta.xpath(
        '//ns:retDistDFeInt/ns:loteDistDFeInt/ns:docZip', namespaces=ns))
    print(f'Quantidade de NSUs na consulta atual: {contador_resposta}')

    cStat = resposta.xpath('//ns:retDistDFeInt/ns:cStat',
                           namespaces=ns)[0].text
    print(f'cStat: {cStat}')

    xMotivo = resposta.xpath(
        '//ns:retDistDFeInt/ns:xMotivo', namespaces=ns)[0].text
    print(f'xMotivo: {xMotivo}')

    maxNSU = resposta.xpath(
        '//ns:retDistDFeInt/ns:maxNSU', namespaces=ns)[0].text
    print(f'maxNSU: {maxNSU}')

    # 137=nao tem mais arquivos e 138=existem mais arquivos para baixar
    if (cStat == '138'):
        for contador_xml in range(contador_resposta):
            tipo_schema = resposta.xpath(
                '//ns:retDistDFeInt/ns:loteDistDFeInt/ns:docZip/@schema', namespaces=ns)[contador_xml]
            numero_nsu = resposta.xpath(
                '//ns:retDistDFeInt/ns:loteDistDFeInt/ns:docZip/@NSU', namespaces=ns)[contador_xml]

            #nfe = 'procNFe_v4.00.xsd'
            #evento = 'procEventoNFe_v1.00.xsd'
            #resumo = 'resNFe_v1.01.xsd'
            if (tipo_schema == 'procNFe_v4.00.xsd'):
                zip_resposta = resposta.xpath(
                    '//ns:retDistDFeInt/ns:loteDistDFeInt/ns:docZip', namespaces=ns)[contador_xml].text

                descompactar_resposta = DescompactaGzip.descompacta(
                    zip_resposta)
                texto_descompactado = etree.tostring(
                    descompactar_resposta).decode('utf-8')

                with open(f'./xml/consulta_distrib-nsu-{NSU}-contador-{contador_xml}.xml', 'w+', encoding='UTF-8') as f:
                    f.write(texto_descompactado)

            # baixar o resumo e realizar a manifestacao do destinatario
            elif (tipo_schema == 'resNFe_v1.01.xsd'):
                zip_resposta = resposta.xpath(
                    '//ns:retDistDFeInt/ns:loteDistDFeInt/ns:docZip', namespaces=ns)[contador_xml].text

                # XML completo do resumo
                resposta_descompactado = DescompactaGzip.descompacta(
                    zip_resposta)

                # Ler chave de acesso do resumo
                chave_acesso_nfe = resposta_descompactado.xpath(
                    '//ns:resNFe/ns:chNFe', namespaces=ns)[0].text

                # Gerar ciÃ¯Â¿Â½ncia da operaÃ¯Â¿Â½Ã¯Â¿Â½o
                # https://github.com/TadaSoftware/PyNFe/wiki/Manifesta%C3%A7%C3%A3o-Destinat%C3%A1rio

        NSU = resposta.xpath(
            '//ns:retDistDFeInt/ns:ultNSU', namespaces=ns)[0].text
        print(f'NSU: {NSU}')

    elif (cStat == '137'):
        print(f'NÃ¯Â¿Â½o hÃ¯Â¿Â½ mais documentos a pesquisar')
        break
    else:
        print(f'Falha')
        break

**Reply by leogregianin** [reply]
@diogoheck6,

Um esclarecimento inicial, não existe nenhum webservice público que faz a baixa dos documentos de saída de NFe, é obrigação do emitente guardá-los. Esse webservice de distribuição não baixa nada de NFC-e (nem entrada e nem saída), para NFCe não tem nenhum webservice de consulta desse tipo.

Fiz uma alteração na parte da gravação do XML em disco e incluí [aqui](https://github.com/TadaSoftware/PyNFe/wiki/Fluxo-da-Consulta-de-Distribuicao-por-NSU) o exemplo do fluxo, compare com seu código e veja onde ocorreu o erro.

**Reply by ibemoreira** [reply]
> Pois é Leo..eu copiei este codigo do git de vocês para tentar conseguir baixar os XML de entrada e saida de NFE e NFCE, pena não ter uma rotina para os de CTE. Pois deixa eu te explicar...hoje eu pego e baixo os extratos do site manualmente e largo numa pasta onde ai sim aplico uma rotina que faz com que se cruze as informações dos extratos com o SPED, gerando um excel com varias informações de quais notas estão no sped e nao estao nos extratos, etc...e Seria um grande saldo se eu conseguisse pegar essas informaçoes via webservice...porém nao entendo quase nada de webservice...até fui olhar as notas técnicas do sefaz para saber por onde começar mas não estou conseguindo..eu programo bastante em python...vou mostrar esse codigo inteiro que te mandei só um pedaço ..se tu puder tentar identificar esse erro..também achei estranho que ele aponta a linha 28 .. O certificado vou colocar como xxxxx por privacidade do cliente..LGPD..se tu puder me dar uma ajuda..ficaria muito feliz mesmo...
> 
> from pynfe.processamento.comunicacao import ComunicacaoSefaz from pynfe.utils.descompactar import DescompactaGzip from pynfe.utils.flags import NAMESPACE_NFE from lxml import etree
> 
> certificado = 'xxxxxxxxxxxxx.pfx' senha = '146531' uf = 'rs' homologacao = True
> 
> CNPJ = '' CHAVE = '' # deixar a chave de acesso vazia NSU = 0
> 
> con = ComunicacaoSefaz(uf, certificado, senha, homologacao) ultNSU = 0 maxNSU = 0 cStat = 0
> 
> while True: xml = con.consulta_distribuicao(cnpj=CNPJ, chave=CHAVE, nsu=NSU) NSU = str(NSU).zfill(15) print(f'Nova consulta a partir do NSU: {NSU}')
> 
> ```
> with open(f'./xml/consulta_distrib_gzip-{NSU}.xml', 'w+') as f:
>     f.write(xml.text)
> 
> # resposta = etree.fromstring(xml.content)
> # print(resposta)
> resposta = etree.parse(f'./xml/consulta_distrib_gzip-{NSU}.xml')
> ns = {'ns': NAMESPACE_NFE}
> 
> contador_resposta = len(resposta.xpath(
>     '//ns:retDistDFeInt/ns:loteDistDFeInt/ns:docZip', namespaces=ns))
> print(f'Quantidade de NSUs na consulta atual: {contador_resposta}')
> 
> cStat = resposta.xpath('//ns:retDistDFeInt/ns:cStat',
>                        namespaces=ns)[0].text
> print(f'cStat: {cStat}')
> 
> xMotivo = resposta.xpath(
>     '//ns:retDistDFeInt/ns:xMotivo', namespaces=ns)[0].text
> print(f'xMotivo: {xMotivo}')
> 
> maxNSU = resposta.xpath(
>     '//ns:retDistDFeInt/ns:maxNSU', namespaces=ns)[0].text
> print(f'maxNSU: {maxNSU}')
> 
> # 137=nao tem mais arquivos e 138=existem mais arquivos para baixar
> if (cStat == '138'):
>     for contador_xml in range(contador_resposta):
>         tipo_schema = resposta.xpath(
>             '//ns:retDistDFeInt/ns:loteDistDFeInt/ns:docZip/@schema', namespaces=ns)[contador_xml]
>         numero_nsu = resposta.xpath(
>             '//ns:retDistDFeInt/ns:loteDistDFeInt/ns:docZip/@NSU', namespaces=ns)[contador_xml]
> 
>         #nfe = 'procNFe_v4.00.xsd'
>         #evento = 'procEventoNFe_v1.00.xsd'
>         #resumo = 'resNFe_v1.01.xsd'
>         if (tipo_schema == 'procNFe_v4.00.xsd'):
>             zip_resposta = resposta.xpath(
>                 '//ns:retDistDFeInt/ns:loteDistDFeInt/ns:docZip', namespaces=ns)[contador_xml].text
> 
>             descompactar_resposta = DescompactaGzip.descompacta(
>                 zip_resposta)
>             texto_descompactado = etree.tostring(
>                 descompactar_resposta).decode('utf-8')
> 
>             with open(f'./xml/consulta_distrib-nsu-{NSU}-contador-{contador_xml}.xml', 'w+', encoding='UTF-8') as f:
>                 f.write(texto_descompactado)
> 
>         # baixar o resumo e realizar a manifestacao do destinatario
>         elif (tipo_schema == 'resNFe_v1.01.xsd'):
>             zip_resposta = resposta.xpath(
>                 '//ns:retDistDFeInt/ns:loteDistDFeInt/ns:docZip', namespaces=ns)[contador_xml].text
> 
>             # XML completo do resumo
>             resposta_descompactado = DescompactaGzip.descompacta(
>                 zip_resposta)
> 
>             # Ler chave de acesso do resumo
>             chave_acesso_nfe = resposta_descompactado.xpath(
>                 '//ns:resNFe/ns:chNFe', namespaces=ns)[0].text
> 
>             # Gerar ciÃ¯Â¿Â½ncia da operaÃ¯Â¿Â½Ã¯Â¿Â½o
>             # https://github.com/TadaSoftware/PyNFe/wiki/Manifesta%C3%A7%C3%A3o-Destinat%C3%A1rio
> 
>     NSU = resposta.xpath(
>         '//ns:retDistDFeInt/ns:ultNSU', namespaces=ns)[0].text
>     print(f'NSU: {NSU}')
> 
> elif (cStat == '137'):
>     print(f'NÃ¯Â¿Â½o hÃ¯Â¿Â½ mais documentos a pesquisar')
>     break
> else:
>     print(f'Falha')
>     break
> ```

@diogoheck6  Eu consegui baixar no XML de NFe emitida contra nosso CNPJ esta rodando liso demais, salvando em uma pasta em nosso servidor usando essa lib, não sou programador estou longe disso mas posso te ajudar se quiser! 

Estou a semanas lutando para tentar fazer uma engenharia reversa em fazer o mesmo para CT-e esta foda mas n vou desistir o @leogregianin  já me deu o caminho das pedra e se você quiser ajudar, estou procurando parceiria!

**Reply by ibemoreira** [reply]
ACABEI DE VER AQUI SAIU ALGO AQUI... VOU TESTAR!

**Reply by diogoheck6** [reply]
@ibemoreira Claro, aceito sua ajuda sim..se quiser anotar meu contato de whats 55 98408-9885 , ai poderiamos conversar melhor e trocar ideias...forte abraço!

**Reply by herbertsmatheus** [reply]
@ibemoreira eu também desejo baixar os XMLs emitidos contra nosso CNPJ, você poderia me ajudar?

Atualmente estou obtendo o seguinte erro:

Description: HTTP 404. The resource you are looking for (or one of its dependencies) could have been removed, had its name changed, or is temporarily unavailable.  Please review the following URL and make sure that it is spelled correctly. 

Requested URL: /NFeDistribuicaoDFe/NFeDistribuicaoDFe.asmx

**Comment by ibemoreira** [comment]
Bom dia, Primeiramente muito obrigado @leogregianin pelo nova implementação passei o final de semana testando, porém parece que ele fica chamando varias vezes o webserver e recebo um kick ban.

`PS C:\Maxvinil\Projeto\PROTHEUS_SQL> & C:/Users/Lucas/AppData/Local/Programs/Python/Python310/python.exe c:/Maxvinil/Projeto/PROTHEUS_SQL/Wilbert/NFeMax/disponi/BAIXAR_CTE.PY
<?xml version="1.0" encoding="UTF-8"?><soap:Envelope xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:xsd="http://www.w3.org/2001/XMLSchema" xmlns:soap="http://www.w3.org/2003/05/soap-envelope"><soap:Header><cteCabecMsg xmlns="http://www.portalfiscal.inf.br/cte/wsdl/CTeDistribuicaoDFe"><cUF>43</cUF><versaoDados>3.00</versaoDados></cteCabecMsg></soap:Header><soap:Body><cteDistDFeInteresse xmlns="http://www.portalfiscal.inf.br/cte/wsdl/CTeDistribuicaoDFe"><cteDadosMsg><distDFeInt versao="1.00" xmlns="http://www.portalfiscal.inf.br/cte"><tpAmb>1</tpAmb><cUFAutor>43</cUFAutor><CNPJ>41591720000156</CNPJ><distNSU><ultNSU>000000000000000</ultNSU></distNSU></distDFeInt></cteDadosMsg></cteDistDFeInteresse></soap:Body></soap:Envelope>
--------------------
C:\Users\Lucas\AppData\Local\Programs\Python\Python310\lib\site-packages\urllib3\connectionpool.py:1043: InsecureRequestWarning: Unverified HTTPS request is being made to host 'www1.cte.fazenda.gov.br'. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/1.26.x/advanced-usage.html#ssl-warnings
  warnings.warn(
Nova consulta a partir do NSU: 000000000000000
Quantidade de NSUs na consulta atual: 50
cStat: 138
xMotivo: documento localizado.
maxNSU: 000000000000196
NSU: 000000000000154
<?xml version="1.0" encoding="UTF-8"?><soap:Envelope xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:xsd="http://www.w3.org/2001/XMLSchema" xmlns:soap="http://www.w3.org/2003/05/soap-envelope"><soap:Header><cteCabecMsg xmlns="http://www.portalfiscal.inf.br/cte/wsdl/CTeDistribuicaoDFe"><cUF>43</cUF><versaoDados>3.00</versaoDados></cteCabecMsg></soap:Header><soap:Body><cteDistDFeInteresse xmlns="http://www.portalfiscal.inf.br/cte/wsdl/CTeDistribuicaoDFe"><cteDadosMsg><distDFeInt versao="1.00" xmlns="http://www.portalfiscal.inf.br/cte"><tpAmb>1</tpAmb><cUFAutor>43</cUFAutor><CNPJ>41591720000156</CNPJ><distNSU><ultNSU>000000000000154</ultNSU></distNSU></distDFeInt></cteDadosMsg></cteDistDFeInteresse></soap:Body></soap:Envelope>
--------------------
C:\Users\Lucas\AppData\Local\Programs\Python\Python310\lib\site-packages\urllib3\connectionpool.py:1043: InsecureRequestWarning: Unverified HTTPS request is being made to host 'www1.cte.fazenda.gov.br'. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/1.26.x/advanced-usage.html#ssl-warnings
  warnings.warn(
Nova consulta a partir do NSU: 000000000000196
Quantidade de NSUs na consulta atual: 0
cStat: 656
xMotivo: Rejeicao: Consumo indevido
maxNSU: 000000000000000
Falha`

**Reply by leogregianin** [reply]
@ibemoreira, o consumo indevido ocorre quando há um grande volume de consultas no servidor da Sefaz, feitas por um mesmo CNPJ e em um curto espaço de tempo. De acordo com as últimas regras publicadas em Notas Técnicas é considerado um uso impróprio do sistema de consultas quando:
 * Não há mais documentos a serem distribuídos e usuário continua consultando;
 * O usuário não está consultando os NSUs (Números Sequenciais Únicos das NFes) de forma sequencial;

**Reply by ibemoreira** [reply]
@leogregianin  Consegui resolver eu estava chamando o WS varias vezes, Obrigado.

**Comment by leogregianin** [comment]
@diogoheck6, para melhor organização das discussões do projeto, pode marcar essa discussão por respondida?
